### PR TITLE
[ORION-3710]: Save and restore named collections

### PR DESCRIPTION
**Pull request by ianton-ru**

### CODE REVIEW

File: ch_backup/logic/named_collections.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+        logging.debug("Performing named collections backup for: {}", " ,".join(nc))
+        for nc_name in nc:
+            context.backup_layout.upload_named_collections_create_statement(
+                context.backup_meta.name, nc_name
+            )
+
+    def restore(self, context: BackupContext) -> None:
+        """
+        Restore UDF objects.
```
Comment (ekopanev) [inline ch_backup/logic/named_collections.py:34]: UDf -> named collections

**Review by ekopanev — CHANGES_REQUESTED** [review]

File: ch_backup/logic/named_collections.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+class NamedCollectionsBackup(BackupManager):
+    """
+    Named collections backup class
+    """
+
+    def backup(self, context: BackupContext) -> None:
+        """
+        Backup named collections.
+        """
+        if not context.ch_ctl.ch_version_ge("23.2"):
```
Comment (MikhailBurdukov) [inline ch_backup/logic/named_collections.py:20]: If ch_version is less than `23.2` then we quit silently? Maybe to put some log msg here?

Comment (ianton-ru) [reply ch_backup/logic/named_collections.py]: Done. And actually method works from 22.12, in 23.2 only was added grants for user.

File: ch_backup/logic/named_collections.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                context.ch_ctl.restore_named_collection(statement)
+
+            logging.debug("Named collection {} restored", nc_name)
+
+        logging.info("All named collections restored")
+
+    @staticmethod
+    def get_named_collections_list(context: BackupContext) -> List[str]:
+        """
+        Get UDF list
```
[the remark this answers is not in the document]
Comment (ianton-ru) [reply ch_backup/logic/named_collections.py]: Fixed

File: ch_backup/logic/named_collections.py (hunk 1 above)
Comment (ianton-ru) [reply ch_backup/logic/named_collections.py]: Fixed

File: ch_backup/ch_backup.py [hunk 4]
```diff
@@ -404,6 +408,12 @@ def restore_access_control(self, backup_name: str) -> None:
         with self._context.locker():
             self._access_backup_manager.restore(self._context)
 
+    def restore_named_collections(self, backup_name: str) -> None:
```
[the remark this answers is not in the document]
Comment (ianton-ru) [reply ch_backup/ch_backup.py]: Removed.

File: tests/integration/modules/clickhouse.py [hunk 5]
```diff
@@ -199,6 +205,13 @@ def get_all_access_objects(self) -> set:
 
         return result
 
+    def get_all_named_collections(self) -> dict:
+        """
+        Retrieve DDL for users, roles, etc
```
Comment (ekopanev) [inline tests/integration/modules/clickhouse.py:210]: "users, roles, etc" -> "named collections"

**Review by ekopanev — APPROVED** [review]
the rest looks good for me, and probably we can change this comment later

File: tests/integration/modules/clickhouse.py (hunk 5 above)
Comment (ianton-ru) [reply tests/integration/modules/clickhouse.py]: Fixed

**Pull request merged**